src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java

```diff
@@ -2924,4 +2947,237 @@ public String toString() {
             return first + ":" + second;
         }
     }
+
+    /**
+     * Return the set of preview language features used to declare the given element.
+     *
+     * @param e the Element to check.
+     * @return the set of preview language features used to declare the given element
+     */
+    @SuppressWarnings("preview")
+    public Set<DeclarationPreviewLanguageFeatures> previewLanguageFeaturesUsed(Element e) {
+        Set<DeclarationPreviewLanguageFeatures> result = new HashSet<>();
+
+        if (e.getKind() == ElementKind.RECORD) {
+            result.add(DeclarationPreviewLanguageFeatures.RECORD);
+        }
+
+        if ((e.getKind().isClass() || e.getKind().isInterface()) &&
+            e.getModifiers().contains(Modifier.SEALED)) {
+            List<? extends TypeMirror> permits = ((TypeElement) e).getPermittedSubclasses();
+            boolean hasLinkablePermits = permits.stream()
+                                                .anyMatch(t -> isLinkable(asTypeElement(t)));
+            if (hasLinkablePermits) {
+                result.add(DeclarationPreviewLanguageFeatures.SEALED_PERMITS);
+            } else {
+                result.add(DeclarationPreviewLanguageFeatures.SEALED);
+            }
+        }
+
+        return result;
+    }
+
+    public enum DeclarationPreviewLanguageFeatures {
```

Comment:
General thinking aloud question ... how does all this interact with source or release options for an earlier release?

Reply:
I don't think there should be much interaction with -source <older-version>. We don't support preview features from previous version or preview class files from previous versions, so I think it should be enough to handle the currently present preview features.

Reply:
We don't support preview features from previous releases, AFAIK. So javadoc for JDK 16 should not accept e.g. record class when running with  -source 15.

Reply:
Yeah, my recollection is that I was wondering whether preview-related code needs to be "guarded" to only work in the current release. But, I guess we may get the right effect (of forbidding preview features in older code) from the javac front end, so that in javadoc we can be assured that there are no instances of what may still be preview features in older code (i.e with older --source/--rlease options)